### PR TITLE
Sanity checks for USB detach result

### PR DESCRIPTION
**Pull request by IvanSanchez**
This is meant to throw an error in any cases where the target devkit does **not** reboot after a detach request. There are two main cases:

- The USB DFU detach request does **not** stall (as expected)
- The `usb` module does **not** fire a `detach` event with the expected device

### CODE REVIEW

File: src/dfuTrigger.js [hunk 1]
```diff
@@ -138,12 +138,26 @@ const getDfuInfo = openDecorator((usbdev, interfaceNumber) => (
...
 const sendDetachRequest = openDecorator((usbdev, interfaceNumber) => (
-    new Promise(resolve => {
+    new Promise((resolve, reject) => {
         usbdev.controlTransfer(
             ReqTypeOUT, DFU_DETACH_REQUEST, 0, interfaceNumber, detachReqBuf,
-            () => {
-                // Device is expected to disappear, any further calls would fail
-                resolve();
+            (err, data) => {
+                // If the detach is sucessfull, the target device will reboot
```
Comment (bihanssen) [inline src/dfuTrigger.js]: /successfull/successful/

File: src/dfuTrigger.js [hunk 2]
```diff
@@ -138,12 +138,26 @@ const getDfuInfo = openDecorator((usbdev, interfaceNumber) => (
...
+                    err.message === 'LIBUSB_TRANSFER_STALL') {
+                    resolve();
+                } else if (err &&
+                    err.errno === usb.LIBUSB_ERROR_IO &&
+                    err.message === 'LIBUSB_ERROR_IO') {
+                    // This edge case only happens when using the "libusb" kernel
+                    // driver on win32 (not "winusb", not "libusbk")
+                    resolve();
+                } else {
+                    reject(new Error('USB DFU detach request sent, but device does not seem to have rebooted'));
```
Comment (mrodem) [inline src/dfuTrigger.js]: Could the control transfer fail with some other code than `LIBUSB_TRANSFER_STALL` or `LIBUSB_ERROR_IO`? In that case, perhaps include the `errno`/`message` in the error message here?

File: src/dfuTrigger.js [hunk 3]
```diff
@@ -165,13 +165,27 @@ const getDfuInfo = openDecorator((usbdev, interfaceNumber) => (
...
             ReqTypeOUT, DFU_DETACH_REQUEST, 0, interfaceNumber, detachReqBuf,
-            () => {
-                // Device is expected to disappear, any further calls would fail
-                resolve();
+            (err, data) => {
+                // If the detach is successful, the target device will reboot
+                // before sending a response, so the expected result is that
+                // the control transfer will stall.
+                if (err &&
+                    err.errno === usb.LIBUSB_TRANSFER_STALL &&
```
Comment (kenr) [inline src/dfuTrigger.js]: Why is it necessary to check both err.errno and err.message?

Comment (IvanSanchez) [reply src/dfuTrigger.js]: It's not necessary, maybe I'm being overcautious here. `errno` should be either a number from [`enum libusb_transfer_status`](http://libusb.sourceforge.net/api-1.0/group__asyncio.html#ga9fcb2aa23d342060ebda1d0cf7478856) or a negative number from [enum libusb_error](http://libusb.sourceforge.net/api-1.0/group__libusb__misc.html#gab2323aa0f04bc22038e7e1740b2f29ef) (in the JS bindings, anything non-zero gets converted into an error).

It *should* be safe to have only one check, if you want.

**Pull request merged**